### PR TITLE
Input x gradient

### PR DESCRIPTION
**Pull request by fanny-jourdan**
## Description

Add input x gradient like a parameter for all gradient based attribution methods.
Add VarGrad, SquareGrad and GradientShap methods
Add tests for aggregation part in attribution methods
Update readme, and docs
Factorize IG perturbator with GradientShap perturbator
Resolve issues #4

## Type of Change

- 📚 docs and tests update
- 🥂 Improvement (non-breaking change which improves an existing feature)
- 🚀 New feature (non-breaking change which adds functionality)

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/docs/contributing.md) guide.
- [x] I've successfully run the style checks using `make lint`.
- [x] I've written tests for all new methods and classes that I created and successfully ran `make test`.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -15,13 +15,16 @@ nav:
...
+        - IntegratedGradients: api/attributions/methods/integrated_gradients.md
         - KernelShap: api/attributions/methods/kernelshap.md
         - Lime: api/attributions/methods/lime.md
-        - IntegratedGradients: api/attributions/methods/integrated_gradients.md
         - Occlusion: api/attributions/methods/occlusion.md
         - Saliency: api/attributions/methods/saliency.md
         - SmoothGrad: api/attributions/methods/smoothgrad.md
+        - SquareGrad: api/attributions/methods/squaregrad.md
         - Sobol: api/attributions/methods/sobol.md
+        - VarGrad: api/attributions/methods/vargad.md
```
Comment (AntoninPoche) [inline mkdocs.yml]: The `VarGrad` page does not exist. You specified the wrong path because of a missing "r":

`.../vargad.md` -> `.../vargrad.md`

File: docs/api/attributions/overview.md [hunk 2]
```diff
@@ -21,9 +21,10 @@ The API have two steps:
...
+For gradient-based attribution methods, granularity is not calculated at the perturbation mask level, but at the aggregation level. We then aggregate the scores given on the tokens to create scores on the chosen granularity. There may be several ways of aggregating these scores. We can also, in the context of gradient-based methods, have a parameter to choose whether we take the gradients directly or the gradients multiplied by the inputs (inspired by the InputxGradient method). So we need new arguments, only for gradient-based methods:
 
 - `granularity_aggregation_strategy` (`GranularityAggregationStrategy`): If granularity is set to 'WORD', this determines how to aggregate token-level attributions into word-level scores. Options are: 'MEAN', 'MAX', 'MIN', 'SUM' and 'SIGNED_MAX'.
+- `input_x_gradient` (bool, optional): If True, multiplies the input embeddings with their gradients before aggregation. Defaults to ``False``.
```
Comment (AntoninPoche) [inline docs/api/attributions/overview.md:27]: Several people, through discussions at ACL, do this by default for gradients. Apparently, there are some mathematical justifications.

Is it a good idea to set it to true by default?

In any case, I would put an example in the attribution walkthrough notebook.

File: docs/roadmap.md [hunk 3]
```diff
@@ -10,25 +10,14 @@ Welcome to the roadmap of Interpreto 🪄. This document outlines the planned fe
...
-
-- **Token-level attribution**
-- **Word-level attribution**
-
-We currently have these attribution methods only with the ALL_TOKENS granularity, which corresponds to having an attribution score for each token in the sentence (even special tokens). We want to align gradient-based methods with inference-based ones.
-
-### 3. Integration of VarGrad
+### 2. Integration of VarGrad
 
 We will implement **VarGrad**, a variance-based extension of gradient attribution. This method improves robustness by computing the **variance of attributions across multiple noisy forward passes**, capturing uncertainty and stabilizing explanations.
```
Comment (AntoninPoche) [inline docs/roadmap.md]: It seems that `VarGrad` is included in your PR. So, this roadmap point is actually outdated.

File: interpreto/attributions/methods/gradient_shap.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    guarantees, while capturing non-linear effects.
+
+    **Reference:**
+    Lundberg and Lee (2017). *A Unified Approach to Interpreting Model Predictions.*
+    [Paper](https://arxiv.org/abs/1705.07874)
+
+    Examples:
+        >>> from interpreto import GradientShap
+        >>> method = GradientShap(model, tokenizer, batch_size=4,
+        >>>                       n_perturbations=20,
```
Comment (AntoninPoche) [inline interpreto/attributions/methods/gradient_shap.py:58]: The `n_perturbations` argument used in your example does not exist. You defined it as `n_interpolations` in the following lines.

Nonetheless, I find `n_perturbations` more natural, even for `SmoothGrad`, `SquareGrad`, and `VarGrad`.

`n_interpolations` was named as such for `IntegratedGradients`. However, I would prefer that we unify all argument names behind `n_perturbations`. Because, even for `IntegratedGradients`, at some point, we perturb inputs, even if this perturbation is an interpolation.

I leave you the final word on this ;)

File: interpreto/attributions/methods/square_grad.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    """
+    SquareGrad is a gradient-based attribution method that computes the variance of input gradients
+    under random perturbations. Unlike methods that average gradients (e.g., SmoothGrad),
+    SquareGrad averages the square of the gradient.
+
+    Procedure:
+    - Generate multiple perturbed versions of the input by adding noise (Gaussian) to the input embeddings.
+    - For each noisy input, compute the gradient of the output with respect to the embeddings.
+    - Average the square of the gradients across all samples.
+    - Aggregate the result per token (e.g., by norm with the input) to get the final attribution scores.
```
Comment (AntoninPoche) [inline interpreto/attributions/methods/square_grad.py:54]: This bullet point list appears as a single line in the documentation.

You should skip a line between each for it to appear correctly in the documentation.

File: interpreto/attributions/methods/var_grad.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    to local perturbations in input space.
+
+    The resulting attributions reveal regions where the gradient signal is consistently volatile,
+    thus potentially highlighting areas where explanations may be less reliable or more fragile.
+
+    Procedure:
+    - Generate multiple perturbed versions of the input by adding noise (Gaussian) to the input embeddings.
+    - For each noisy input, compute the gradient of the output with respect to the embeddings.
+    - Compute the element-wise variance of the gradient values across these samples.
+    - Aggregate the result per token (e.g., by norm with the input) to get the final attribution scores
```
Comment (AntoninPoche) [inline interpreto/attributions/methods/var_grad.py]: Samer list-documenation problem as `GradientShap`. You have to skip a line between each point.

File: interpreto/attributions/perturbations/gradient_shap_perturbation.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    def __init__(self, inputs_embedder=None, baseline=None, n_perturbations=10, std=0.1):
+        """
+        Initializes the GradientShapPerturbator.
+
+        Args:
+            inputs_embedder: Optional module to transform inputs into embeddings.
+            baseline: The reference embedding (can be a tensor, float, int, or None).
+            n_perturbations: Number of random samples for interpolation.
+            std: Standard deviation of the Gaussian noise added to the baseline.
```
Comment (AntoninPoche) [inline interpreto/attributions/perturbations/gradient_shap_perturbation.py]: No type specified.

File: interpreto/attributions/perturbations/linear_interpolation_perturbation.py [hunk 8]
```diff
@@ -47,8 +48,10 @@ def __init__(
         Initializes the LinearInterpolationPerturbation instance.
 
         Args:
+            inputs_embedder: Optional module to transform inputs into embeddings.
             baseline (TensorBaseline, optional): The baseline value for the perturbation.
                 It can be a torch.Tensor, int, float, or None. Defaults to None.
+            n_perturbations: Number of interpolation steps between baseline and input.
```
Comment (AntoninPoche) [inline interpreto/attributions/perturbations/linear_interpolation_perturbation.py]: Even in the `LinearInterpolationPerturbator` it is called `n_perturbations`. I reaffirm that the `n_interpolations` could be replaced by `n_perturbations` in the library.

File: interpreto/attributions/methods/gradient_shap.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        model: PreTrainedModel,
+        tokenizer: PreTrainedTokenizer,
+        batch_size: int = 4,
+        granularity: Granularity = Granularity.WORD,
+        granularity_aggregation_strategy: GranularityAggregationStrategy = GranularityAggregationStrategy.MEAN,
+        device: torch.device | None = None,
+        inference_mode: Callable[[torch.Tensor], torch.Tensor] = InferenceModes.LOGITS,
+        input_x_gradient: bool = False,
+        n_interpolations: int = 10,
+        baseline: torch.Tensor | float | None = None,
```
Comment (AntoninPoche) [inline interpreto/attributions/methods/gradient_shap.py:75]: You also support `int` here. A type `TensorBaseline` was created, but I am unsure if it is useful. Sometimes, leaving all the possibilities readable is easier to follow.

File: interpreto/attributions/perturbations/linear_interpolation_perturbation.py [hunk 10]
```diff
@@ -91,19 +94,40 @@ def adjust_baseline(baseline: TensorBaseline, inputs: torch.Tensor) -> torch.Ten
...
+        # Repeat and flatten the attention mask accordingly: (b, l) -> (n*b, l)
+        attn = model_inputs["attention_mask"]  # (b, l)
+        attn = attn.unsqueeze(0).repeat(self.n_perturbations, 1, 1).reshape(self.n_perturbations * b, -1)
+        model_inputs["attention_mask"] = attn
 
-        model_inputs["inputs_embeds"] = (1 - alphas) * embeddings + alphas * baseline
-        model_inputs["attention_mask"] = model_inputs["attention_mask"].repeat(
-            model_inputs["inputs_embeds"].shape[0], 1
-        )
         return model_inputs, None
```
Comment (AntoninPoche) [inline interpreto/attributions/perturbations/linear_interpolation_perturbation.py]: To my understanding, we only perturb on single inputs. Hence, the n in your code is 1. You could simplify your code and the comments a bit.

Other remarks:
- What you call `b` is called `p` for perturbations in the rest of the code.
- You used the `jaxtyped` decorator, but did not use jaxtyping inside your function; it could help verify everything is as expected.

File: interpreto/attributions/perturbations/gradient_shap_perturbation.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        - Replicates the baseline for each interpolation step and batch element.
+        - Adds Gaussian noise with standard deviation `std`.
+        """
+        baseline = self.adjust_baseline(self.baseline, embeddings)
+        baseline = baseline.to(embeddings.device)
+        b = embeddings.shape[0]
+
+        baseline = baseline.unsqueeze(0).expand(b, *baseline.shape)  # (b, l, d)
+        baseline = baseline.unsqueeze(0).repeat(self.n_perturbations, 1, 1, 1)  # (n, b, l, d)
+        baseline += torch.randn_like(baseline) * self.std  # bruit
```
Comment (AntoninPoche) [inline interpreto/attributions/perturbations/gradient_shap_perturbation.py]: `bruit` -> `noise`

Comment (AntoninPoche) [inline interpreto/attributions/perturbations/gradient_shap_perturbation.py]: Is `b` the batch_size or the number of perturbations? It seems to be the batch_size, but you use `n` afterward which is usually the batch_size. I am a bit lost.

Furthermore, we perturb input by input; thus, the batch size should always be 1.

File: interpreto/attributions/methods/gradient_shap.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    **Reference:**
+    Lundberg and Lee (2017). *A Unified Approach to Interpreting Model Predictions.*
+    [Paper](https://arxiv.org/abs/1705.07874)
+
+    Examples:
+        >>> from interpreto import GradientShap
+        >>> method = GradientShap(model, tokenizer, batch_size=4,
+        >>>                       n_perturbations=20,
+        >>>                       baseline=0,
+        >>>                       noise_level=0.1,)
```
Comment (AntoninPoche) [inline interpreto/attributions/methods/gradient_shap.py]: Another name I find non-adapted, `noise_level`, is not super informative; I think `noise_std` would be much more informative.

This name appears in several other methods; you could change them all if you think it's relevant.

**Review by AntoninPoche — CHANGES_REQUESTED** [review]
Super clean, thanks a lot @fanny-jourdan !!!

Only minor comments on documentation, argument names, and comments.

File: interpreto/attributions/methods/integrated_gradients.py [hunk 13]
```diff
@@ -70,7 +70,8 @@ def __init__(
         granularity_aggregation_strategy: GranularityAggregationStrategy = GranularityAggregationStrategy.MEAN,
         device: torch.device | None = None,
         inference_mode: Callable[[torch.Tensor], torch.Tensor] = InferenceModes.LOGITS,
-        n_interpolations: int = 10,
+        input_x_gradient: bool = False,
```
Comment (AntoninPoche) [inline interpreto/attributions/methods/integrated_gradients.py]: You changed the default value of `input_x_gradient` to `True` in the five other methods. But forgot `IntegratedGradients`.

There are two modifications to do: the attributes and the docstring.

File: interpreto/attributions/base.py [hunk 14]
```diff
@@ -163,6 +163,7 @@ def __init__(
         granularity_aggregation_strategy: GranularityAggregationStrategy = GranularityAggregationStrategy.MEAN,
         inference_mode: Callable[[torch.Tensor], torch.Tensor] = InferenceModes.LOGITS,  # TODO: add to all classes
         use_gradient: bool = False,
+        input_x_gradient: bool = False,
```
Comment (AntoninPoche) [inline interpreto/attributions/base.py]: You changed the value of `input_x_gradient` to `True` in the methods, but did not update the base class.

There is the attribute and the docstring to update.

**Review by AntoninPoche — APPROVED** [review]
Well done!

You just missed to classes when changing `input_x_gradient` default value.
